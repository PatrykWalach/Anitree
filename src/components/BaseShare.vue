<template>
  <v-dialog v-model="isShared" scrollable max-width="440px">
    <v-card>
      <v-tooltip bottom>
        <template v-slot:activator="{ on, attrs }">
          <v-btn
            :style="{ top: 0, right: 0 }"
            absolute
            icon
            small
            v-bind="attrs"
            @click="isShared = false"
            v-on="on"
          >
            <v-icon>close</v-icon>
          </v-btn>
        </template>
        <span>Close</span>
      </v-tooltip>
      <v-card-title>Share link</v-card-title>
      <v-card-text>
        <v-slide-group show-arrows>
          <base-share-item
            :href="`http://www.facebook.com/sharer.php?u=${url}`"
            color="#3B5998"
            title="Facebook"
          >
            <FontAwesomeIcon :icon="['fab', 'facebook-f']" />
          </base-share-item>
          <base-share-item
            :href="`https://twitter.com/intent/tweet?url=${url}`"
            color="#1DA1F2"
            title="Twitter"
          >
            <FontAwesomeIcon :icon="['fab', 'twitter']" />
          </base-share-item>

          <base-share-item
            color="#35465C"
            :href="`http://www.tumblr.com/share/link?url=${url}`"
            title="Tumblr"
          >
            <FontAwesomeIcon :icon="['fab', 'tumblr']" />
          </base-share-item>
          <base-share-item
            :href="`http://www.reddit.com/submit?url=${url}`"
            color="#FF4500"
            title="Reddit"
          >
            <FontAwesomeIcon :icon="['fab', 'reddit-alien']" />
          </base-share-item>
          <base-share-item :href="`mailto:?body=${url}`" title="E-mail">
            <v-icon dark>email</v-icon>
          </base-share-item>
        </v-slide-group>

        <v-snackbar v-model="snackbar" left>
          Link copied to clipboard
          <v-btn color="accent" text @click="snackbar = false">
            Close
          </v-btn>
        </v-snackbar>

        <v-card flat>
          <v-list-item v-clipboard="url" @click.stop @success="snackbar = true">
            <v-list-item-content>
              <v-list-item-subtitle>
                <v-text-field
                  hide-details
                  flat
                  single-line
                  solo
                  readonly
                  :value="url"
                ></v-text-field>
              </v-list-item-subtitle>
            </v-list-item-content>
            <v-list-item-action>
              <v-btn>
                copy
              </v-btn>
            </v-list-item-action>
          </v-list-item>
        </v-card>
      </v-card-text>
    </v-card>
  </v-dialog>
</template>
<script lang="ts">
import { library } from '@fortawesome/fontawesome-svg-core'
import {
  faFacebookF,
  faTwitter,
  faRedditAlien,
  faTumblr
} from '@fortawesome/free-brands-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import BaseShareItem from './BaseShareItem.vue'
library.add(faFacebookF, faTwitter, faRedditAlien, faTumblr)
import { clipboard } from 'vue-clipboards'

import { ShareData } from '../types'
import { createComponent, computed, ref } from 'vue-function-api'
import useShare from '../store/share'
interface Props {
  options: ShareData | null
}
export default createComponent({
  directives: {
    clipboard
  },
  components: { FontAwesomeIcon, BaseShareItem },
  setup(props: Readonly<Props>) {
    const snackbar = ref(false)
    const { isShared } = useShare()

    const url = computed(() => (props.options && props.options.url) || '')

    return { isShared, url, snackbar }
  },
  props: {
    options: { required: true }
  }
})
</script>
