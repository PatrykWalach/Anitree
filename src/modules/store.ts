import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
