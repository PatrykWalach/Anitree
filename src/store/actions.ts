export { titleActions as title } from './modules/title'

export { editActions as edit } from './modules/edit'
export { filterActions as filter } from './modules/filter'
export { shareActions as share } from './modules/share'
export { settingsActions as settings } from './modules/settings'
export { timelineActions as timeline } from './modules/timeline'
export { changesActions as changes } from './modules/changes'
