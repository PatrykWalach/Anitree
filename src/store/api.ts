import gql from 'graphql-tag'
import { FetchVariables } from './models'

import apollo from '../apollo'
import { AMedia } from '@/types'

export const query = gql`
  query($search: String, $idIn: [Int], $page: Int = 0) {
    Page(page: $page, perPage: 50) {
      media(search: $search, id_in: $idIn) {
        id
        bannerImage
        coverImage {
          extraLarge
          large
          medium
          color
        }
        format
        status
        tags {
          name
          id
          rank
        }
        siteUrl
        type
        studios(isMain: true) {
          nodes {
            name
            id
          }
        }
        startDate {
          year
          month
          day
        }
        endDate {
          year
          month
          day
        }
        title {
          romaji
          english
          native
        }
        description
        mediaListEntry {
          status
        }
        relations {
          edges {
            node {
              id
              type
            }
            id
            relationType
          }
        }
      }
    }
  }
`

export function fetchMediaApollo(variables: FetchVariables): Promise<AMedia[]> {
  return apollo
    .query<{ Page: { media: AMedia[] } }>({
      query,
      variables
    })
    .then(({ data }) => data.Page.media)
}
