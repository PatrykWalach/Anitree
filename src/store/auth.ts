import VueFunctionApi, { ref, Ref, computed } from 'vue-function-api'
import Vue from 'vue'
Vue.use(VueFunctionApi)

import { Token } from '@/types'
const stored = localStorage.getItem('AUTH')

const auth: Ref<Token | null> = ref((stored && JSON.parse(stored)) || null)

const token = computed(() => {
  if (
    auth.value
    // && Date.now() < parseInt(auth.value.expires_in)
  ) {
    return auth.value.access_token
  }
  return null
})

const CHANGE_TOKEN = async (payload: Token | null) => {
  if (payload) {
    const newAuth: Token = {
      ...payload,
      /* eslint-disable-next-line @typescript-eslint/camelcase */
      expires_in: (Date.now() + parseInt(payload.expires_in)).toString()
    }
    localStorage.setItem('AUTH', JSON.stringify(newAuth))
    auth.value = newAuth
  } else {
    auth.value = payload
  }
}

export default function useAuth() {
  return { auth, CHANGE_TOKEN, token, stored }
}
