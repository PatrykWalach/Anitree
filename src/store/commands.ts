import CompositionApi, { Ref, ref, watch } from '@vue/composition-api'
import {
  DeleteCommand,
  DeleteCommandConstructor,
} from './commands/DeleteCommand'
import { DeleteVariables, SaveVariables } from '@/graphql/schema/listEntry'
import { SaveCommand, SaveCommandConstructor } from './commands/SaveCommand'
import { Form } from '@/types'
import { FuzzyDate } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import Vue from 'vue'
import { useSettings } from './settings'

Vue.use(CompositionApi)

export interface Command {
  saveState?(): void | Promise<void>
  undo?(): void | Promise<void>
  execute(): void | Promise<void>
}

export interface ListCommand extends Command {
  backup: Form | null
  variables: SaveVariables | (DeleteVariables & Pick<SaveVariables, 'mediaId'>)
  done: boolean
  _class: 'SaveCommand' | 'DeleteCommand'
}

export const isDeleteCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is DeleteCommandConstructor & Pick<DeleteCommand, '_class'> =>
  e._class === 'DeleteCommand'

export const isSaveCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is SaveCommandConstructor & Pick<SaveCommand, '_class'> =>
  e._class === 'SaveCommand'

const stored = localStorage.getItem('CHANGES')

const history: Ref<ListCommand[]> = ref(
  (stored &&
    JSON.parse(stored, (key, value) => {
      if (value instanceof Object) {
        if (isSaveCommand(value)) {
          return new SaveCommand(value)
        }
        if (isDeleteCommand(value)) {
          return new DeleteCommand(value)
        }
      }
      return value
    })) ||
    [],
)

watch(() => {
  if (useSettings().cacheChanges.value) {
    localStorage.setItem('CHANGES', JSON.stringify(history.value))
  }
})
;(async () => {
  for (const command of history.value.filter(({ done }) => !done)) {
    await command.execute()
  }
})()

const dispatch = async (command: ListCommand) => {
  history.value.push(command)
  await command.execute()
  return command
}

export const useCommands = () => {
  return {
    dispatch,
    history,
  }
}

export const mediaListToForm = (
  mediaListEntry: MediaList | null,
  advancedScoring: string[],
): Form => {
  if (mediaListEntry) {
    const advancedScores = advancedScoring
      .map(key => mediaListEntry.advancedScores[key])
      .map(value => value || 0)

    delete mediaListEntry.__typename
    delete mediaListEntry.startedAt.__typename
    delete mediaListEntry.completedAt.__typename
    delete mediaListEntry.private
    delete mediaListEntry.id
    delete mediaListEntry.mediaId

    return {
      ...mediaListEntry,
      advancedScores,
    }
  }

  const advancedScores = advancedScoring.map(() => 0)
  return {
    advancedScores,
    completedAt: { day: null, month: null, year: null },
    notes: '',
    progress: 0,
    progressVolumes: 0,
    repeat: 0,
    score: 0,
    startedAt: { day: null, month: null, year: null },
    status: null,
  }
}

export const isDate = (e: any): e is FuzzyDate => e.__typename === 'FuzzyDate'
