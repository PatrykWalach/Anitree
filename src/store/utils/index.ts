export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
