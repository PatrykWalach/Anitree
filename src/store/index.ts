import Vue from 'vue'
import Vuex from 'vuex'

// import { getModule } from 'vuex-module-decorators'
Vue.use(Vuex)

import { ModuleMedia } from './modules/media'
import { ModuleTheme } from './modules/theme'
import { ModuleTitle } from './modules/title'
import { ModuleAuth } from './modules/auth'

export interface RootState {
  media: ModuleMedia
  theme: ModuleTheme
  title: ModuleTitle
  auth: ModuleAuth
}

export default new Vuex.Store<RootState>({})
