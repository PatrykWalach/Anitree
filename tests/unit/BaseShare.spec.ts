import BaseShare from '@/components/BaseShare.vue'

import { matchSnapshot } from './utils'
import { useMockedStore } from './mocks/store'

describe('BaseShare', () => {
  matchSnapshot(BaseShare, {
    propsData: {
      options: null,
    },
    ...useMockedStore(),
  })
})
