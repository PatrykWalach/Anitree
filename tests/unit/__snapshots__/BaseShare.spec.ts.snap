// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseShare matches snapshot 1`] = `
<v-dialog
  max-width="440px"
  scrollable=""
>
  <v-card-stub
    loaderheight="4"
    tag="div"
  >
    <v-card-title-stub>
      Share link
      
      <v-spacer-stub />
       
      <v-tooltip-stub
        closedelay="0"
        contentclass=""
        fixed="true"
        maxwidth="auto"
        nudgebottom="0"
        nudgeleft="0"
        nudgeright="0"
        nudgetop="0"
        nudgewidth="0"
        opendelay="0"
        openonhover="true"
        tag="span"
        top="true"
      >
         
        <span>
          Close
        </span>
      </v-tooltip-stub>
    </v-card-title-stub>
     
    <v-snackbar-stub
      left="true"
      timeout="6000"
    >
      
      Link copied to clipboard
      
      <v-btn-stub
        activeclass=""
        color="accent"
        tag="button"
        text="true"
        type="button"
      >
        
        Close
      
      </v-btn-stub>
    </v-snackbar-stub>
     
    <v-list-stub
      tag="div"
      tile="true"
    >
      <v-list-item-stub
        activeclass=""
        tag="div"
      >
        <v-list-item-content-stub>
          <v-list-item-subtitle-stub>
            <v-text-field-stub
              backgroundcolor=""
              clearicon="$clear"
              errorcount="1"
              errormessages=""
              flat="true"
              hidedetails="true"
              loaderheight="2"
              messages=""
              readonly="true"
              rules=""
              singleline="true"
              solo="true"
              successmessages=""
              type="text"
              value=""
            />
          </v-list-item-subtitle-stub>
        </v-list-item-content-stub>
         
        <v-list-item-action-stub>
          <v-btn-stub
            activeclass=""
            tag="button"
            type="button"
          >
            
            copy
          
          </v-btn-stub>
        </v-list-item-action-stub>
      </v-list-item-stub>
    </v-list-stub>
     
    <baseshareitems-stub
      url=""
    />
  </v-card-stub>
</v-dialog>
`;
